pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 Chaos Mesh Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/robfig/cron/v3"
+
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+)
+
+const (
+	// Set the top bit if a star was included in the expression.
+	starBit = 1 << 63
+)
+
+// LastTime returns the next time this schedule activated, less than or equal with the given time.
+func LastTime(spec v1alpha1.SchedulerSpec, now time.Time) (*time.Time, error) {
+	scheduler, err := cron.ParseStandard(spec.Cron)
+	if err != nil {
+		return nil, fmt.Errorf("fail to parse runner rule %s, %v", spec.Cron, err)
+	}
+	var next time.Time
+	if cronSpec, ok := scheduler.(*cron.SpecSchedule); ok {
+		scheduleLast := &cusSchedule{cronSpec}
+		next = scheduleLast.Last(now)
+	} else if cronSpec, ok := scheduler.(cron.ConstantDelaySchedule); ok {
+		scheduleLast := &cusConstantDelaySchedule{cronSpec}
+		next = scheduleLast.Last(now)
+	} else {
+		return nil, fmt.Errorf("assert cron spec failed")
+	}
+	return &next, nil
+}
+
+type cusConstantDelaySchedule struct {
+	cron.ConstantDelaySchedule
+}
+
+// Last returns the last time this schedule activated, less than or equal with the given time.
+// So it would always return now
+func (s cusConstantDelaySchedule) Last(t time.Time) time.Time {
+	return t
+}
+
+type cusSchedule struct {
+	*cron.SpecSchedule
+}
+
+// Last returns the last time this schedule activated, less than or equal with the given time.
+// If no time can be found to satisfy the schedule, return the zero time.
+// Modified from the original `Next` function in robfig/cron at Dec 15, 2020
+func (s *cusSchedule) Last(t time.Time) time.Time {
+	// General approach:
+	// For Month, Day, Hour, Minute, Second:
+	// Check if the time value matches.  If yes, continue to the next field.
+	// If the field doesn't match the schedule, then increment the field until it matches.
```

Comment:
Should `increment` change to `decrease`?

Reply:
this function decrease 1 second every time, may loop too much times, is there other better ways to get the last time?

Reply:
> Should increment change to decrease?
Thanks! Fixed.

> this function decrease 1 second every time, may loop too much times, is there other better ways to get the last time?
Implementing `-1s`(😎) is to decrease to the last block of time, say last year. For example, when we are at the first second of 2021, -1s could take us to 2020 but not to miss any seconds in 2020.
Since it's used to decrease block of time, the efficiency is still the same with [function Next in cron pkg](https://github.com/robfig/cron/blob/master/spec.go).